### PR TITLE
Multi-Image IP-Adapter

### PR DESCRIPTION
**Pull request by RyanJDick**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Description

This PR adds the ability to pass multiple images to a single IP-Adapter (note the difference from using _multiple IP-Adapters at once_, which is already supported.). The image embeddings are combined in the IP-Adapter attention layers. This is the same strategy for combining multiple images as used in Insta-LoRA workflows (https://civitai.com/articles/2345).

This PR only adds multi-image support in the backend and the node editor. The Linear UI still needs to be updated.

## QA Instructions, Screenshots, Recordings

I have manually tested the following via the workflow editor:
- Multiple images with a single IP-Adapter
- Multiple images per IP-Adapter, and multiple IP-Adapters
- Both standard and sequential conditioning
- IP-Adapters still work in the Linear UI.

Please hammer at this feature some more with manual testing.

## Added/updated tests?

- [x] Yes
- [ ] No

I updated the existing IP-Adapter smoke test, but it provides pretty limited coverage of this feature. This feature would probably be best tested by an end-to-end workflow test, which is not currently supported. (I'm hoping to put some effort into workflow-level testing soon.)

### CODE REVIEW

**Comment by RyanJDick** [comment]
@hipsterusername This is a different implementation than we had previously discussed. Images are combined in the IP-Adapter attention layer as opposed to combining the CLIP Vision embeddings. This means that there is no weighting of individual images. This is how its done in the Insta-LoRA workflows.

Let me know what you think.

File: invokeai/backend/ip_adapter/attention_processor.py [hunk 1]
```diff
@@ -127,26 +133,35 @@ def __call__(
...
                     query, ip_key, ip_value, attn_mask=None, dropout_p=0.0, is_causal=False
                 )
 
+                # ip_hidden_states.shape: (batch_size, num_heads, query_seq_len, head_dim)
+
                 ip_hidden_states = ip_hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
                 ip_hidden_states = ip_hidden_states.to(query.dtype)
 
+                # ip_hidden_states.shape: (batch_size, query_seq_len, num_heads * head_dim)
+
```
Comment (RyanJDick) [inline invokeai/backend/ip_adapter/attention_processor.py:164]: Note to code reviewers who may be surprised to see no code changes in this block:

I was surprised too. This code actually works with `ipa_embed`s of shape `(batch_size, num_ip_images, ip_seq_len, ip_image_embedding)` OR `(batch_size, ip_seq_len, ip_image_embedding)`. Hopefully the shape comments that I added make it clear how this works.

**Comment by hipsterusername** [comment]
> @hipsterusername This is a different implementation than we had previously discussed. Images are combined in the IP-Adapter attention layer as opposed to combining the CLIP Vision embeddings. This means that there is no weighting of individual images. This is how its done in the Insta-LoRA workflows.
> 
> Let me know what you think.

I think that makes the UX easier, so we’ll go with it :)

**Comment by psychedelicious** [comment]
Need some ideas for UI for this

Separate component for multi image ip adapter?

Link ip adapters and disable the weight sliders when linked?

**Comment by blessedcoolant** [comment]
Umm .. allow adding of multiple images to the IP adapter image upload area and that solves it? Or am I missing something here?

Coz the entire adapter itself has weighting. The only thing that changes is the number of image inputs so all it really needs is just support for adding more than one image to the same adapter.

**Comment by psychedelicious** [comment]
Ya that's the first option separate component that can handle multiple images. But what did the ui for that look like? It's a very small area at have to work with at the moment

Probably we need a separate control adapters panel bc it's such an important area

**Comment by blessedcoolant** [comment]
> Ya that's the first option separate component that can handle multiple images. But what did the ui for that look like? It's a very small area at have to work with at the moment
> 
> Probably we need a separate control adapters panel bc it's such an important area

Maybe coz the IP images are not processed, 1 image full size, 2 images, smaller thubmanail 50/50 .. 3 thumbnails in 1 row max ..and expand as it goes in to the row after? Mostly people wont put more than a couple of images.. so it wont ever grow too far ..?

**Comment by psychedelicious** [comment]
The instant Lora workflow uses like 6 images but dunno if that's really common

Regardless we need a multiple image container relatively soon. For this and nodes. So maybe we can figure one out that works for both

**Comment by blessedcoolant** [comment]
> The instant Lora workflow uses like 6 images but dunno if that's really common
> 
> Regardless we need a multiple image container relatively soon. For this and nodes. So maybe we can figure one out that works for both

6 is fine i think? 2 rows of small thumbnails. .. 3 each? Even 9 would still be about the same size as the current full single image display right?

**Comment by psychedelicious** [comment]
> 6 is fine i think? 2 rows of small thumbnails. .. 3 each? Even 9 would still be about the same size as the current full single image display right?

How to display them isn't the hard part - like a small grid or whatever - what's hard is making a user interface to allow you to interact with up to 6 or 9 different images in that tiny area. Will be a nodes-only thing to start

**Comment by blessedcoolant** [comment]
> > 6 is fine i think? 2 rows of small thumbnails. .. 3 each? Even 9 would still be about the same size as the current full single image display right?
> 
> How to display them isn't the hard part - like a small grid or whatever - what's hard is making a user interface to allow you to interact with up to 6 or 9 different images in that tiny area. Will be a nodes-only thing to start

What interaction elements would the IP adapter images require? Just the ability to remove right?

**Comment by psychedelicious** [comment]
> What interaction elements would the IP adapter images require? Just the ability to remove right?

Add & remove if it was only for that

But we need a component, like IAIDndImage, that encapsulates all the logic related to managing and displaying a collection of images. Should have a thumbnail view and full view. Needs to accept DND of multiple images or single, stuff like that.

File: invokeai/backend/ip_adapter/attention_processor.py [hunk 2]
```diff
@@ -127,26 +133,35 @@ def __call__(
...
                 assert ipa_embed.shape[0] == encoder_hidden_states.shape[0]
-                # The channel dimensions should match.
-                assert ipa_embed.shape[2] == encoder_hidden_states.shape[2]
+                # The token_len dimensions should match.
+                assert ipa_embed.shape[-1] == encoder_hidden_states.shape[-1]
 
                 ip_hidden_states = ipa_embed
 
+                # ip_hidden_state.shape = (batch_size, num_ip_images, ip_seq_len, ip_image_embedding)
+
```
Comment (psychedelicious) [inline invokeai/backend/ip_adapter/attention_processor.py:142]: is this just showing the expected shape or is this something that should be uncommented in the future?

**Comment by hipsterusername** [comment]
> > What interaction elements would the IP adapter images require? Just the ability to remove right?
> 
> Add & remove if it was only for that
> 
> But we need a component, like IAIDndImage, that encapsulates all the logic related to managing and displaying a collection of images. Should have a thumbnail view and full view. Needs to accept DND of multiple images or single, stuff like that.

What if we just have a viewer that allows for toggling between uploaded images but no “gallery”? Effectively, same as current component with a <\> capability on the bottom (and a small number icon) a la staging menu for the small/expanded views?

**Comment by psychedelicious** [comment]
> What if we just have a viewer that allows for toggling between uploaded images but no “gallery”? Effectively, same as current component with a <> capability on the bottom (and a small number icon) a la staging menu for the small/expanded views?

Ya could work, still a lot of buttons in a tiny space but I guess there's no getting around that without creating a separate, larger UI for control adapters

**Review by psychedelicious — APPROVED** [review]
Works fine for me. Tried a selection of drastically different image sizes, if that matters. Not sure how else to pressure test.

File: invokeai/backend/ip_adapter/attention_processor.py (hunk 2 above)
Comment (RyanJDick) [reply invokeai/backend/ip_adapter/attention_processor.py]: This is just the expected shape. I updated the comments in [3079c75](https://github.com/invoke-ai/InvokeAI/pull/4882/commits/3079c75a60c01f5478f3230a8f74159d347fb1a0) to make this more clear.

**Comment by Millu** [comment]
Just gave this a go - I think we're good to merge!

**Review by brandonrising — APPROVED** [review]

**Pull request merged**